### PR TITLE
ci: add workflow to check for vulnerabilities in images

### PR DESCRIPTION
**Pull request by rkpattnaik780**
Add automation to generate a weekly report of security vulnerabilities found by Quay.

## Description

A workflow has been added to display the vulnerabilities found in the different workbench images in tabular format.

## How Has This Been Tested?

A dry run in fork will result the following changes. [link](https://github.com/rkpattnaik780/notebooks/pull/33/files)

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: .github/workflows/sec-scan.yml [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+      - name: install python packages
+        run: |
+          python -m pip install --upgrade pip
+          pip install requests
+
+      - name: execute py script  # run trial.py
+        env:
+          HASH_N: ${{ steps.hash-n.outputs.HASH_N }}
+          RELEASE_VERSION_N: 2023b
```
Comment (atheo89) [inline .github/workflows/sec-scan.yml:109]: Hmm why do you hard code the RELEASE_VERSION_N again here? I see that they are already declared on the 15 and 16 line

File: Makefile [hunk 2]
```diff
@@ -468,4 +468,9 @@ refresh-pipfilelock-files:
...
 	cd runtimes/tensorflow/ubi9-python-3.9 && pipenv lock
 	cd base/c9s-python-3.9 && pipenv lock
-	
\ No newline at end of file
+	
+# This is only for the workflow action
+# For running manually, set the required environment variables
+.PHONY: scan-image-vulnerabilities
+scan-image-vulnerabilities:
+	python ci/security-scan/quay_security_analysis.py
```
Comment (atheo89) [inline Makefile:476]: Could you add a new line under 476 without spaces to remove this red icon?

**Review by atheo89 — CHANGES_REQUESTED** [review]
Great contribution, Rama! I've gone through this pull request and appended a few comments for minor adjustments.

**Comment by atheo89** [comment]
Tnx /lgtm :slightly_smiling_face:

**Comment by jiridanek** [comment]
I have one (additional) suggestion. Enable Dependabot security alerts for this repo.

Currently, GitHub (dependabot) is already scanning dependencies in this repository and lists them on https://github.com/opendatahub-io/notebooks/network/dependencies. What it does not do is show the vulnerablities tab, which must be enabled extra https://docs.github.com/en/code-security/dependabot/dependabot-alerts/about-dependabot-alerts#configuration-of-dependabot-alerts

File: ci/security-scan/quay_security_analysis.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+for i, image in enumerate(IMAGES_MAIN):
+    process_image(image, commit_id_path, "", LATEST_MAIN_COMMIT)
+
+branch_main_data = ""
+for key, value in branch_dictionary.items():
+    branch_main_data += f"| [{key}](https://quay.io/repository/opendatahub/workbench-images/manifest/{branch_dictionary[key]['sha']}?tab=vulnerabilities) |"
+    for severity in ['Medium', 'Low', 'Unknown', 'High', 'Critical']:
+        count = value.get(severity, 0)  # Get count for the severity, default to 0 if not present
+        branch_main_data += f" {count} |"
+    branch_main_data += "\n"
```
Comment (harshad16) [inline ci/security-scan/quay_security_analysis.py:128]: We should move this to a function as it being used multiple times.
```suggestion
def get_branch_data(branch_dictionary):
    branch_data = ""
    for key, value in branch_dictionary.items():
        branch_data += f"| [{key}](https://quay.io/repository/opendatahub/workbench-images/manifest/{branch_dictionary[key]['sha']}?tab=vulnerabilities) |"
        for severity in ['Medium', 'Low', 'Unknown', 'High', 'Critical']:
            # Get count for the severity, default to 0 if not present
            count = value.get(severity, 0)
            branch_data += f" {count} |"
        branch_data += "\n"
    return branch_data
```

**Review by harshad16 — CHANGES_REQUESTED** [review]
Great work.
sorry for the delayed review.

Adding comment on at least to those code block where an new action is happening would help contributor.

File: .github/workflows/sec-scan.yml [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+          pr_body: |
+            :rocket: This is an automated Pull Request.
+
+            This PR updates:
+
+            * `ci/security-scan/security_scan_results.md` file with the latest security vulnerabilities reported by Quay.
+            * `ci/security-scan/weekly_commit_ids` with the latest updated SHA digests of the notebooks (N & N-1)
+            Created by `/.github/workflows/sec-scan.yaml`
+
+            :exclamation: **IMPORTANT NOTE**: Remember to delete the ` ${{ env.SEC_SCAN_BRANCH }}` branch after merging the changes
```
Comment (jiridanek) [inline .github/workflows/sec-scan.yml:148]: Consider enabling https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-the-automatic-deletion-of-branches in repository settings

Comment (rkpattnaik780) [reply .github/workflows/sec-scan.yml:148]: Thanks for the suggestion. We can club it with automating deletion for [this workflow](https://github.com/opendatahub-io/notebooks/blob/main/.github/workflows/notebooks-digest-updater-upstream.yaml#L182) as well.

File: .github/workflows/sec-scan.yml [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+# The aim of this GitHub workflow is to update the `ci/securitty-scan/security_scan_results.md` with latest security scan results.
+name: Update notebook image security reports
+on:
+  workflow_dispatch:
+    inputs:
+      branch:
+        required: true
+        description: "Provide the name of the branch you want to update ex main, vYYYYx etc: "
+  schedule:
+    - cron: "0 0 * * 0"  #Scheduled every Friday
```
Comment (harshad16) [inline .github/workflows/sec-scan.yml:11]: perhaps change this to every sprint.

**Review by harshad16 — APPROVED** [review]
/lgtm
/approve

thanks for the work :+1:

**Pull request merged**